### PR TITLE
Focus on current subtree instead of root

### PR DESCRIPTION
**Pull request by saurfang**
This fixes an issue where ref can be ambiguous when drilling down.

Take the following code as example:

```r
library(lineprof)
library(magrittr)

quick <- function() { pause(.1) }
slow <- function() { pause(.2) }

quickWrap <- . %>% {quick()} 
slowWrap <- . %>% {slow()} 

x <- lineprof({ quickWrap(); slowWrap() })
print(x, depth = 5)
shine(x)
```

The results in the following
```
Reducing depth to 5 (from 7)
   time alloc release dups                                                              ref
1 0.083 0.003       0   18 c("quickWrap", "freduce", "withVisible", "<Anonymous>", "quick")
2 0.161 0.000       0    3   c("slowWrap", "freduce", "withVisible", "<Anonymous>", "slow")
                                              src
1 quickWrap/freduce/withVisible/<Anonymous>/quick
2 slowWrap/freduce/withVisible/<Anonymous>/slow  
```

If we drill down a couple levels to `withVisible` from `slowWrap`, we find 

1. total time is incorrect 
image
image

2. drilling down one level deeper tells us the function uses both `slow` and `quick`. This is incorrect because we drilled down from `slowWrap` thus only `slow` should be shown.
image

This can be fixed by making `focus` use the current `stack$top()` rather than the whole profiling tree `x`.

After the fix, we see things are now correctly identified.
image
image

Please let me know if you have any questions or concerns.

### CODE REVIEW

**Comment by hadley** [comment]
Thanks! Do you see an easy way to add a test for this problem?

**Comment by saurfang** [comment]
AFAIK there isn't a headless testing framework for shiny, this is the best thing I can think of for now. Would be far cooler if we can do headless testing on Shiny (maybe with phamtomjs?)

I only constructed test case for this particular problem. Let me know if you want to change anything.

Couple issues I faced:

1. We can't use `lineprof` directly in tests because that would fail R CMD CHECK. I commented out the code that generates the `test-shine.prof` file. Is there a more appropriate place for me to put it like `data-raw`?
2. I had to wrap everything in `eval(parse(test=...)` otherwise things will be quoted as refs in file instead of function call stacks.
3. I don't like the way I import stuff under `shine` into testthat and would appreciate if you can show me any better ninja tricks.
4. I made a change so that `align` is only called if the path exists. For one thing, the file may no longer exists and things can still work perfectly since at least we know the function call stacks. However, the change is also necessary when profiling code with `parse(text = ...)` where the path will show up as `<text>`.

**Comment by hadley** [comment]
Oh I didn't realise all this code was inside `shine()`. Probably just easier to skip the test for now :(  If instead you wanted to do this "right", you'd modify shine to separate out the shiny code from the stack navigation code so that you could more easily test it.

I'd put the data in the tests directory (since it's used for tests). It's probably better to generate once anyway, since that avoids the chance that the stack isn't as you expect

Change to `align` sounds good to me.

**Comment by saurfang** [comment]
Okay. I followed your suggestion and moved `navigate` out of `shine` function so we can directly access/test it. I also moved the prof file to test directory and attached R script for reproducibility. I think I named it correctly so it doesn't actually get run during testing. Finally I took the liberty to add a few test for the stack function as well while I'm at it.

Hope these should conclude this PR. Let me know if you have any questions or suggestions.

File: tests/testthat/navigate-test.R [hunk 1]
```diff
@@ -0,0 +1,9 @@
+path <- line_profile(eval(parse(text = "
```
Comment (hadley) [inline tests/testthat/navigate-test.R]: Would you mind adding a note about the use of parse? (i.e. needed to get line numbers)

**Comment by hadley** [comment]
One tiny change, otherwise looks great - thanks!

**Comment by saurfang** [comment]
Done. Let me know if you prefer to squash the commits into one.

**Comment by hadley** [comment]
Thanks!

**Comment by saurfang** [comment]
Thank you for the great package!

**Pull request merged**